harness/determined/keras/_tf_keras_trial.py

```diff
@@ -459,67 +538,100 @@ def run(self) -> None:
                 path = cast(pathlib.Path, args[0])
                 response_func(self._save_checkpoint(path))
             elif wkld.kind == workload.Workload.Kind.TERMINATE:
-                self.model.stop_training = True
-                self.expect_terminate = True
                 response_func({} if self.is_chief else workload.Skipped())
-                break
+                self.multiplexer.on_train_end()
+                self._stop_training_check()
+                raise det.errors.WorkerFinishedGracefully
             else:
-                raise AssertionError(f"Unknown wkld kind {wkld.kind}.")
+                raise AssertionError(f"Unknown workload kind {wkld.kind}.")
 
-    def _launch_fit(self) -> None:
-        check.false(self.fit_loop_started)
-        self.fit_loop_started = True
+    def _allreduce_logs(self, logs: Dict) -> Dict:
+        if not self.hvd_config.use:
+            return logs
+        return {key: np.array(hvd.allreduce(metric)) for key, metric in logs.items()}
+
+    def _post_train_batch_end(self, num_inputs: int, logs: Dict) -> None:
+        # Remove default keras metrics we aren't interested in like "batch" and "size".
+        self.train_workload_metrics.append(
+            {k: v for k, v in logs.items() if k not in {"batch", "size"}}
+        )
+        self.train_workload_inputs += num_inputs
+        self.train_workload_batches += 1
 
-        self.tf_keras_callbacks.append(DeterminedEarlyStoppingCallback(self))
-        self.tf_keras_callbacks.append(WaitForInstructionsCallback(self))
+        if self.train_workload_batches != self.train_workload_len:
+            return
+
+        if self.train_response_func is None:
+            raise AssertionError(
+                "Callback should avoid calling model.predict(), "
+                "as this will affect Determined training behavior",
+            )
 
         if self.hvd_config.use:
-            # When using horovod broadcast initial variable states from rank 0 to
-            # all other processes.
-            self.tf_keras_callbacks.append(hvd.callbacks.BroadcastGlobalVariablesCallback(0))
+            num_inputs = hvd.allreduce(num_inputs, average=False)
 
-        (
-            training_input,
-            batches_per_epoch,
-        ) = self._train_input_manager.get_training_input_and_batches_per_epoch()
+        # Return only the latest metrics.
```

Comment:
non-blocking: may be useful to include that we do this because keras keeps a running average.